### PR TITLE
Containers: reuse HTTP connections more.

### PR DESCRIPTION
**Pull request by tmds**
## Description

Closes https://github.com/dotnet/sdk/issues/35753

This removes the artificially-short Pool Lifetime used by the HttpClientHandler that the SDK Container publishing tooling uses to communicate with remote container registries.
This short lifetime was something we put in very early on in the lifetime of the code, and likely was part of a mitigation around our incorrect authentication handling in earlier versions.

## Customer Impact

This short lifetime lead to many connections getting used instead of fewer connections being re-used. Some registries may have configurations that rely on some connection-local state (like sticky sessions) and using more connections that required can impact those flows. In addition, reusing connections improves performance, because negotiating new connections takes non-trivial amounts of time.

## Regression

**No** - we've had this code since our early versions.

## Risk

**Low** - while this does change the configuration of the HttpClientHandler used for all registry communication, testing against our set of supported live registries shows no regression in behaviors, and manual testing confirms this.

### CODE REVIEW

**Comment by tmds** [comment]
@baronfel I've targeted the 8.0 branch for now. I can change to the main branch is that is preferable.

@antonfirsov if you want, can you take a look at these HttpClient related changes and approve the PR if they look good to you?

**Comment by baronfel** [comment]
This one almost certainly will not make RC2. We could consider for GA, though - I'll try to kick off a run of sdk-container-demo with this change to help serve as evidence that things still behave.

File: src/Containers/Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs
Comment (tmds) [reply src/Containers/Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs:329]: I think the difference is that here we're in a `DelegatingHandler` so we don't get the `HttpCompletionOption` behavior.

The connection becomes available when the response content gets read entirely.
It doesn't require the response object to be disposed.

I'm reading the content so the object can still be returned from the method.

Comment (baronfel) [reply src/Containers/Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs:329]: Ok, that fills in the gap for me. Thank you!

**Review by baronfel — APPROVED** [review]

**Comment by baronfel** [comment]
Here's an Actions run using this payload: https://github.com/baronfel/sdk-container-demo/actions/runs/6382668229/job/17321761464

It's looking green as expected. Manually checking requests using a local HTTP inpector looks fine as well.

File: src/Containers/Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs
Comment (antonfirsov) [inline src/Containers/Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs:329]: Nit: is it really important to have the buffering cancellable? How much data is drained typically? Cancellation will close and drop the connection. Is this something you want here?

https://github.com/dotnet/runtime/blob/82fee2692b3954ba8903fa4764f1f4e36a26341a/src/libraries/System.Net.Http/src/System/Net/Http/HttpContent.cs#L473-L475

**Comment by baronfel** [comment]
Approved over email - @tmds can you check out @antonfirsov's review comment above and respond before this is merged?

File: src/Containers/Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs
Comment (tmds) [reply src/Containers/Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs:329]: > Is it really important to have the buffering cancellable

It's important we drain the response while respecting the cancellation token.

> How much data is drained typically?

Not much. None to a few hundred of bytes at most.

> Cancellation will close and drop the connection. Is this something you want here?

Yes.

**Review by antonfirsov — APPROVED** [review]

File: src/Containers/Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs
Comment (danmoseley) [inline src/Containers/Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs:328]: > LoadIntoBufferAsync, but it has no overload that accepts a CancellationToken

worth an API proposal?

**Pull request merged**
